src/vsr/client.zig

```diff
@@ -296,210 +210,80 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
         }
 
-        /// After writing Events to batch.slice(), mark the batched data as sendable with a callback
-        /// and user_data for when it completes. This may also start sending the Batch's message
-        /// through VSR.
-        pub fn batch_submit(
-            self: *Self,
-            user_data: u128,
-            callback: Request.Callback,
-            batch: Batch,
-        ) void {
-            assert(batch.request.demux_queue.count > 0);
-            assert(batch.request.message.body().len > 0);
-            assert(batch.request.message.header.command == .request);
-
-            batch.demux.user_data = user_data;
-            batch.demux.callback = callback;
-
-            // Newly made Requests are populated and potentially pushed through VSR if there exists
-            // no currently inflight request.
-            // A Batch is newly made if its Request only contains its Demux.
-            if (batch.request.demux_queue.count == 1) {
-                assert(batch.request.demux_queue.peek() == batch.demux);
-                self.request_submit(batch.request);
-            }
-        }
-
-        /// Sends a request, only setting request_number in the header. Currently only used by
-        /// AOF replay support to replay messages with timestamps.
+        /// Sends a request, only setting request_number in the header.
+        /// There must be no other request message currently inflight.
         pub fn raw_request(
             self: *Self,
-            user_data: u128,
             callback: Request.Callback,
+            user_data: u128,
             message: *Message.Request,
         ) void {
-            assert(!message.header.operation.vsr_reserved());
-            assert(self.messages_available < constants.client_request_queue_max);
-
-            const event_size: usize = switch (message.header.operation.cast(StateMachine)) {
-                inline else => |operation| @sizeOf(StateMachine.Event(operation)),
-            };
-
-            // Assume that the caller ensures there will be Demux nodes available for the message.
-            const demux = self.demux_pool.acquire() orelse unreachable;
-            errdefer self.demux_pool.release(demux);
-
-            demux.* = .{
-                .user_data = user_data,
-                .callback = callback,
-                .event_count = if (event_size == 0) 0 else @intCast(
-                    @divExact(message.body().len, event_size),
-                ),
-                .event_offset = 0,
-            };
+            assert(message.header.client == self.id);
+            assert(message.header.release.value == self.release.value);
+            assert(message.header.cluster == self.cluster);
+            assert(message.header.command == .request);
+            assert(message.header.size >= @sizeOf(Header));
+            assert(message.header.size <= constants.message_size_max);
+            assert(message.header.operation.valid(StateMachine));
```

Comment:
Can we assert that `message.header.timestamp == 0 or constants.aof_recovery);`?

Reply:
This [was reached](<https://github.com/tigerbeetle/tigerbeetle/actions/runs/8364883560/job/22901324354?pr=1640>) under the existing `test_aof` step so I'll comment it out for now.